cleaner.go

```diff
@@ -0,0 +1,362 @@
+	pullCount := 0
+	flushPulls := func() error {
+		result, err := pullChunk.Run()
+		if err != nil {
+			if err != mgo.ErrNotFound {
+				// not found is odd, but not considered fatal, all others are
+				return fmt.Errorf("error while updating documents: %v", err)
+			}
+		}
+		cleaner.stats.UpdatedDocCount += result.Matched
+		pullChunk = cleaner.config.Source.Bulk()
+		pullChunk.Unordered()
+		pullCount = 0
+		return nil
+	}
+	for _, doc := range cleaner.docsToProcess {
+		toPull := cleaner.findPullableTokens(doc.Queue, completedTokens)
+		if cleaner.removeIfEmpty && len(toPull) == len(doc.Queue) {
+			// this document can just be removed from the stash
+			cleaner.docIdsToRemove = append(cleaner.docIdsToRemove, doc.Id)
+		} else if len(toPull) > 0 {
+			// Note: (jam 2017-04-04) An observation, if it is legal to pull a token from one document,
+			// it is legal to pull it from all other documents. We could do bulk operations by using the
+			// union of all document ids and the union of all tokens to pull.
+			cleaner.stats.PulledTokenCount += len(toPull)
+			pull := bson.M{"$pullAll": bson.M{"txn-queue": toPull}}
+			pullChunk.Update(bson.M{"_id": doc.Id}, pull)
+			if pullCount >= maxBulkOps {
+				if err := flushPulls(); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	if err := flushPulls(); err != nil {
+		return err
+	}
+	// We've handled these tokens and documents
+	cleaner.tokensToLookup = cleaner.tokensToLookup[:0]
+	cleaner.docsToProcess = cleaner.docsToProcess[:0]
+	return nil
+}
+
+// flushRemoveQueue ensures that all pending removals are flushed to the database.
+func (cleaner *collectionCleaner) flushRemoveQueue() error {
+	if len(cleaner.docIdsToRemove) == 0 {
+		return nil
+	}
+	remover := newBulkRemover(cleaner.config.Source)
+	for _, docId := range cleaner.docIdsToRemove {
+		if err := remover.remove(docId); err != nil {
+			return fmt.Errorf("failed while removing document %v from %q", docId, cleaner.config.Source.Name)
+		}
+	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("failed while removing documents from %q", cleaner.config.Source.Name)
+	}
+	cleaner.stats.RemovedCount += remover.removed
+	logger.Debugf("flushing %d documents removed %d (%d total)",
+		len(cleaner.docIdsToRemove), remover.removed, cleaner.stats.RemovedCount)
+	cleaner.docIdsToRemove = cleaner.docIdsToRemove[:0]
+	return nil
+}
+
+// Cleanup iterates the collection and ensures that all documents no longer reference completed transactions.
+func (cleaner *collectionCleaner) Cleanup() error {
+	// walk through all the documents and remove what we can, update what we cannot
+	logger.Debugf("cleaning up completed documents from %q", cleaner.config.Source.Name)
+	// If we delete documents while we iterate, it can cause the iterator to miss documents.
+	// So we do multiple passes on the database to make sure we catch everything.
+	var doc txnDocument
+	for iterCount := 0; iterCount < maxIterCount; iterCount++ {
```

Comment:
Just thinking about this some more... is doing multiple passes over a collection really better using a temporary working collection? This approach seems to have introduced an awful lot of complexity, not too mention the non-trivial memory usage.

On a busy live system I could imagine this looping up to maxIterCount on a number of collections which isn't ideal.